Multiple same-type database functionality

panosfol opened the pull request:
Closes #63

panosfol committed:
Initial commit

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
Initial config support & argument parsing

Config fields added:
       Postgres
	 host     : string
	 user     : string
	 password : string
	 port     : string
	 database : string

Arguments added:
       "-f" | "--config-file": Config file to use

We also revisited our Makefile in this commit to avoid the following
issues:
   * object files not being found, undefined functions were
   being called due to linking errors
   * Added DEBUG conditional flag for debug printing
   * Made CC a user-specified variable (if not specified, use gcc)
   * Add -lcyaml
   * Added valgrind

It is also safe to say that you need to install `libcyaml` for this
patch to work on your computer.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
config: add "enabled" postgres config field

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
main: move free() calls outside if statement

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
Add `connect to postgres` functionality

This was achieved by utilizing the `libpq-fe` library.
The `construct_pg_info` function using the parsed `yaml`
file is called to initialize the `values` array
that will be passed to `PQconnectdbParams` function.
Finally the connection is checked and then closed.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
Add scripts for creating Postgres databases using Docker

There are 2 scripts, 1 for deploying 1 for destroying
the Postgres container.
This is mainly for testing purposes and use basic/default
values for configurating the databases.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
Makefile: use correct libpq-dev include location

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
Cast `values` array to `const char const *` when connecting
to Postgres

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
config: do not use pointer to bool for "enabled"

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
db: introduce db_operations mechanism

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
db: postgres: use db_operations mechanism

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
Makefile: do not use LDFLAGS in object build step

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
Adding `pg_connect` utilizing the db-operations API

This commit also contains some minor memory fixes.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
log: introduct "-v | --verbose" flag and simple logger

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
makefile: remove redundant -DDEBUG compile flag

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
config: add origin_ and target_ postgres config fields

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
Merge remote-tracking branch 'origin/master' into db-ops-api

charmitro committed:
log: use "get_verbose()" in pr_info macro

Without the get_verbose() function, all invocations of "pr_info" in a
file that is not "log.c" will need "extern bool verbose;"to work.

This fixes this issue by calling "get_verbose()" right from "log.c"
where "verbose" is initial defined.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
Add `replicate` functionality for Postgres and minor fixes

The main logic of `replicate` is achieved by using `fork`
to create two new processes, one for `pg_dump` and one for
`pg_restore`. The commit also contains utilization of `log.h`
library and some fixes for better error handling.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
scripts: make postgres-{up|down}.sh executable for all

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
log: rename "pr_info" to "pr_debug"

"pr_info" can be confused with the programs regular STDOUT print
function. Renaming it to "pr_debug" makes more sense since we only
print when on verbose mode.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
config: migrate to ini.h from cyaml

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
config: introduce smtp config section

We are preparing the ground for adding SMTP email support.

Added in this commit:
      1. SMTP config section (`smtp_config`)
      	 * username
	 * password
	 * smtp_host
	 * smtp_port
	 * auth_mode (TLS || SSL)
      2. Macro CHECK_SECTION
      	 In order to avoid checking all fields and waste time, we are
	 checking if the section is the one we want and then fall into
	 the relevant if-else clauses.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

Commit:
db: postgres: simplify replication process

Signed-off-by: Panagiotis Foliadis <pfoliadis@Debian-bookworm-latest-amd64-base>

Commit:
Fix memory leak

Signed-off-by: Panagiotis Foliadis <pfoliadis@Debian-bookworm-latest-amd64-base>

panosfol committed:
Add new README.adoc file

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

Commit:
Add error handling for ini_parse and handler functions of inih library
Fix memory leak by calling `free_config` before returning in main.c

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
add email functionality

1. The Makefile has been updated to link the curl library, which is
necessary for the new email functionality.

2. A new `email.adoc` documentation file has been added. This file
provides detailed information about the `send_email` function and the
global variables used in the email sending process.

These changes are part of an effort to enhance the project's
communication capabilities by enabling it to send emails.

Also fixed:
	* Wrongly returning in main() after config_file != NULL

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
docs: add docs

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
Fix returning on `initialize_config` error

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
Github CI actions for Valgrind

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
Add GPLv3 License

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
config: fix `ini_parse` argument & add config for GH actions

- Fix `ini_parse` to use `config_file` argument instead of
having the file name hardcoded.
- Add custom config to enable Github Actions.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
config: fix mem-leak on config file was not-found

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
docs: add db-op-api & config documentation

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
Update README.adoc

charmitro committed:
Update README.adoc

charmitro committed:
Update README.adoc

charmitro committed:
ci: add clang-format test

Also
	* Fixed some clang-format errors:
		* include/email.h
		* include/db/postgres.h
	* Changed job name from
		"runner-job" -> "build-and-valgrind"

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
db-ops-api.adoc: cleanup & add more info

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
main: add help flag and message

Usage:
	cnc -h
	cnc --help

Help message is also triggered by not providing any arguments.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
main: resolve segfault when an unknown non-option argument is given

Previous:
	$ ./cnc dsadsa
	Segmentation fault

Current:
	$ ./cnc dsadsa
	Non-option argument: dsadsa
	Usage: cnc [options]
	Options:
	  -f, --config-file <file>  Specify the config file to use
	  -v, --verbose             Run in verbose mode
	  -h, --help                Print this help message

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
Makefile: fix valgrind command

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
tests: introduce tests

Moved CNC's main function logic to seperate function in cnc.h/cnc.c to
allow us to use it in tests.

We're using CUnit for our testing suite.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
postgres: add backup-type field & minor fixes

This commit contains the following changes:
- config.h: add a new field to `postgres_t`

- postgres.c: change the postgres-replicate function to provide two
different kinds of `pg_dump`, `schema-only` and `full`

- Move the enum `backup_type` from `db/db.h` to `config.h`,
and remove the `struct_options` completely

- config.c: fix `if` statement in the `smtp` section

- config.c: add an `if` statement to check for acceptable values for
fields `backup_typ` and `auth_mode` of the `.ini` file

- Add test cases for backup type

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
Add Docker support and documentation for CNC application

This commit introduces Docker support for the CNC application. It
includes a Dockerfile for building a lightweight, Alpine-based Docker
image for the application. The Dockerfile is set up with a multi-stage
build process: the first stage compiles the application using Alpine
as the base image and the second stage prepares the final image
containing only the necessary runtime dependencies and the compiled
binary.

Additionally, a documentation file (docker.adoc) is added. Providing
instructions on how to build and run the CNC application inside a
Docker container, including how to mount a configuration
file from the host into the container.

Changes include:
	- Dockerfile: Defines the multi-stage build process for creating a
	  minimal Docker image.
	- docs/docker.adoc: Contains detailed instructions for
	  building the Docker image and running the CNC application
	  inside a Docker container.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
README: reference our new docker docs

charmitro committed:
readme: mark emailing as under development

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
rust: add email shared library

Changelog:
	* Adjust our Makefile to build libraries under `rust/`
	* Add `email` shared library using cbindgen
	* Call the function to test
	* Add from, to, cc config variables (single strings for now)
	* Allow errors of valgrind in CI

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
transition project build system to automake

This commit introduces the automake build system to the project,
replacing the manually written Makefile with a more flexible and
portable automake configuration. This change enables better handling
of different environments and simplifies the build process for
contributors and users.

Changes include:
    - Added `configure.ac` with checks for essential libraries (inih,
      libpq, libcurl) and header files, along with adjustments for
      macOS Homebrew paths for PostgreSQL and other dependencies.
    - Created `Makefile.am` specifying the sources, flags, and targets
      for the main executable and tests, ensuring proper compilation and
      linkage.
    - Implemented `autogen.sh` to automate the generation of the
      `configure` script and `Makefile.in` files, streamlining the
      initial setup process.
    - Updated `.gitignore` to exclude auto-generated files from
      automake, keeping the repository clean.
    - Removed the old `Makefile` as it is now generated by automake,
      ensuring that the build process is fully managed by the new
      system.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
db: postgres: fix compiler warnings

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
rust: enhance email functionality

This update significantly enhances the interface for sending emails
from C to Rust by introducing a structured approach with the
`EmailInfo` struct. The `EmailInfo` struct formalizes the parameters
for sending an email, including sender, recipients, CC, and body,
making it easier for C developers to utilize the Rust email sending
functionality.

Key changes include:

  - Introduction of the `EmailInfo` struct in C bindings, providing a
    clear and structured way to pass email information from C to Rust.
  - Detailed documentation comments added to `email-bindings.h` to
    guide C developers on using the `EmailInfo` struct and the
    `send_email` function, with example usage that demonstrates
    constructing the `EmailInfo` struct and invoking the email sending
    process.
  - Refinement of the Rust `send_email` function to accept an
    `EmailInfo` struct as its parameter, streamlining the process of
    email construction and sending within the Rust code.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
postgres: fix path strings length

This was observed:
   $ PG_BIN=/usr/bin/ ./cnc -f configs/test.ini -v
   configs/test.ini
   VERBOSE:connect_pg	Origin-database connection: Success!
   VERBOSE:connect_pg	Target-database connection: Success!
   VERBOSE:replicate	Checking if $PG_BIN environmental variable.
   VERBOSE:replicate	$PG_BIN=/usr/bin/ was found.
   Error executing `pg_dump`
   VERBOSE:execve_binary	`pg_dump` error: No such file or directory
   VERBOSE:execve_binary	`pg_dump` was unsuccessful. Replication process will be terminated.
   `pg_dump` failed

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
log: add pr_{info | error | warn} log macros

Also replaced all printf's/fprintf's with the new functions.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
rust: email: do not panic if `cc` is null

Also changed tabs to spaces in doc comments, as clippy suggested.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
config: smtp: SMTP configuration to support multiple recipients

Changelog:
  - The `test.ini` file has been updated to include examples of multiple
    recipients for both `to` and `cc` fields.
  - Adjusted the `smtp_t` structure in `include/config.h` to accommodate
    arrays for `to` and `cc`, allowing up to 10 recipients initially for
    each field.
  - Introduced utility functions `remove_spaces` and `split_str` in
    `src/config.c` to parse the comma-separated email addresses from the
    configuration file and prepare them for use in the SMTP
    settings.
  - Replacing the previous single-string assignment with a more robust
    handling suitable for multiple recipients.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
rust: email: support multiple CC recipients

Changelog:
  - Modifying the `EmailInfo` struct to support an array of strings for
    both `to` and `cc` fields, including the length of these arrays.
  - Updating the example usage documentation within `email-bindings.h`
    to reflect the ability to specify multiple `to` and `cc` recipients.
  - Simplifying the email sending function `send_email` in `lib.rs` by
    updating the handling of CC recipients to iterate over an array of
    strings.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
config: add `enabled` field for [smtp] & `email` field for [postgres]

This commit contains changes to add and parse the new field,
which must both be true in order for the email functionality
to be used for a given database system.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
rust: change subject: "Calling Rust from C" -> "CNC digest"

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
postgres: Add email functionality

This commit contains the necessary changes to send an email
of the result of a database replication:

- add `email_body` variable to send the results of `replicate`

- add error handling for `read_buffer_pipe` function

- Add `src/util.c` & `include/util.h` that contain
  general-purpose code that will be used by other Database systems

- read_buffer_pipe(), execve_binary() definitions
  were moved to util.h and declarations to util.c

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
build-system: remove libcurl

Since the migration of the email implementation from C -> Rust,
libcurl is no longer needed.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
cnc: catch if config_file is NULL

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
config: fix memory leaks in config string handling and cleanup

String handling in config parsing to prevent memory leaks by
duplicating strings and properly freeing resources.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
util.h: add macro for allocating memory

The macro was created to pair together allocating memory
and checking if the memory allocated is NULL.
All the existing malloc() calls where replaced by the macro
and error handling was added.

Test cases were also added, and the `Makefile.am` file was
modified to ignore some expected warnings only for the tests.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
Add `d_str_t` struct to `postgres_t`

The addition of this struct was made to pair up
all the same fields in the `postgres_t` struct from the
origin and target database in a single unit.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
config: log: add `.log` file for log keeping

This commit contains the necessary changes for log keeping as well as
a new section [general] for the `.ini` file with `log_filepath` as a new field.

Changelog:

config.h: add [general] section and `log_filepath` field
	& a new struct to hold the value of the `.ini` file

config.c: add parsing of the new section

log.h:
	- add a new macro similar to pr_{debug|info|error|warn} that writes
	- both to the log file and the STDOUT/STDERR
	- add macro for `LOG_FILEPATH` and `LOG_FILENAME`
	- add `construct_log_filepath` function, which is responsible for
	creating and checking the necessary permissions for log-keeping
	directory.
	- add `construct_log_filename` function that creates the name of the
	log file with the format `cncDDMMYY_HHMMSS.log`.

util.h:
	- add `mkdir_p` function to create iteratively the parent directories
	of a user defined path.
	- add `cnc_strdup` function that calls `strdup` and checks if the call
	was successful.

config.c: change `initialize_config` to construct the log file utilizing
	the new functions added to `log.h`

cnc.c: add error handling if `initialize_config` fails to open the log file

cnc_tests.c: add 2 tests for the new field

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
postgres: util: db: append the log file with the output

Utilize the `log` interface to create and append to a log file the
results of the `pg_replicate`.

postgres.c: util.c: utilize the new macro for appending to the log file
		    remove the usage of variables around constructing
		    an email body

postgres.h: add `PG_DUMP_FILE` macro and change the declaration
	of `construct_pg` function to return an error code in
	case of not being able to open the `.log` file

util.h: remove the unecessary arguments for the email body and size

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
fix compiler warning

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
fix: pr_info_fd macro printing two times on stdout

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

zycon91 committed:
rust: read email body from file

This commit will make it possible to read the mail body from a file. The `EmailInfo` struct
will have a `filepath` parameter instead of `body`.

Changelog:
	- Changed `EmailInfo` parameter from `body` to `filepath`
	- Added a helper function `read_email_body()` to read the context of the file located in `filapath`
Signed-off-by: Fotis Zikos <fzikos@gmail.com>

panosfol committed:
postgres: add email functionality

Utilize the email functionality from rust to send the result
of the replicate over email

Changelog:
	- `stdrup` the fully constructed filename to
	ini_config->general_config->filepath
	- create the `email_info` struct and send_email in `close_pg`
	- call `fseek` before and after sending email to correctly
	place the file position	indicator

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
util: remove redundant free(..)

If "strdup(..)" fails to allocate, it returns "NULL". Calling "free()"
is no-op (i.e. it does nothing).

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
util: make "cnc_strdup(..)" inline

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
db: remove struct d_str_t declaration (duplicate)

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
db: automake: resolve macOS build system issue

We had problems building/running on macOS due to the differences between
GCC and Clang compilers. Clang could not treat our ADD_FUNC macro suite
the way we intented to do.

Apart from that we also had problems including error.h and errno.h.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
rust: email: cargo update

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
utils: log: config: make some utility functions inline static

This commit inlines several small utility functions, namely
`remove_spaces`, `get_verbose`, and `cnc_strdup`, directly within
their respective header files to potentially reduce function call
overhead and improve performance. By making these functions inline, we
hint to the compiler that it should attempt to embed these functions
directly at the call site, which can lead to faster execution for
these frequently used utilities.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
rust: email: cargo update

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
cnc: Add --version flag to display version and commit hash

This patch introduces a new `--version` command-line flag to the cnc
program, allowing users to easily retrieve the current version number
and the Git commit hash associated with the build. When invoked with
`--version` or `-V`, the program will display this information,
providing valuable insights into the exact version and build being
used.

More,
  * Update configure.ac to dynamically set the version using `git describe`
  * Add COMMIT_HASH macro in Makefile.am and configure.ac
  * Modify cnc.c to handle the --version flag and display version information
  * Update help message to include the --version flag

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
util: move `construct_dump_path` function to util and rename
to `construct_filepath`

This choice was made since many different database systems
would use this function to construct the path required for
dumping their respective file.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
util: add macro for max length of a filepath

The length was chose according to the Linux/Windows
specification and is 256 characters. This addition
makes it more clear which variables are holding filenames.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
postgres.c: remove the `.dump` file that was created by backup

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
Makefile.am: Add COMMIT_HASH build flag while building tests

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
db: postgres: fix flow of execution

This commit solves a problem which arose when the main
logic of the database replication was prematurely exited
if one of the database implementations was disabled or
was unable to establish a connection. Also there was a problem
with initializing the `available_dbs` struct with multiple
database systems

Changelog:
db.c: - add a global variable counter to accurately track the
	initialization of the `available_dbs` struct
      - change the core loop to not return when a database
	fails to connect or is disabled

postgres.c: - add a check to see if database is enabled or not
	    - utilize the new global variable to populate the
	      `available_dbs` struct

postgres.h: move the documentation for the `connect_pg` function
	    from `postgres.c` to `postgres.h`

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>
Co-Developed-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
cnc: fix potential segfault if we encounter parsing error

Calling `pr_error_fd` when we encounter a parsing error leads
to seg fault since we haven't actually opened the stream to write
to the log file.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
log: free `log_filepath`

The `log_filepath` which was allocated on config.c
is being overwritten in `construct_log_filepath` and that
results in memory leak. Freeing before overwritting solves
that problem.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
config: postgres: add multiple database functionality

This commit contains the necessary changes to implement
a multiple same-database system, utilizing a linked-list.
Each linked list represents a database system, with each
node being a different database.

Changelog:
	- For each section parsed in the config file, create
	  a node. If the config contains multiple of the same
	  section, i.e. multiple databases of the same type,
	  then allocate memory for a new node that will hold
	  the config for each database.
	- Changed `db_t` struct to take a `void *` argument
	  instead of a specific database structure, and also
	  change some functions in a similar way. This change
	  was necessary in order to leverage generic programming
	  for the main functions (connect, replicate, close).
	- Added some helper function for the linked list.
	- Added more thread safe functionality to critical sections
	  of the codebase.
	- Changed the structure of the `configs` file, added `.ini`
	  files for testing single/multple databases

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>
Co-developed-by: Charalampos Mitrodimas <charmitro@posteo.net>

Commit:
db: postgres: Migrate pg_dump to pg_dumpall && pg_restore to psql

By using "pg_dumpall" instead of "pg_dump" we ensure that ALL
resources of the database are being exported in the backup file, which
is now an SQL script instead of `.dump` due to the nature of
"pg_dumpall". This means that previous issues are now resolved,
i.e.
	* Missing privileges
	* Missing roles

We also migated from "pg_restore" to the standard "psql" command,
running it in the format of,
	psql -h <address> -U <user> -p <port> -f <SQL script>

This, with the addition of "pg_dumpall" solves our major issue of
being unable to run "pg_restore" on an already "restored" database.

Signed-off-by: Charalampos Mitrodimas <charmiro@posteo.net>

charmitro committed:
docker: update Dockerfile to use Debian

Image size: 192MB.

Use `sh <script>` in Makefile.am to fix issue with docker running the
build-rust-libs.sh script.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

charmitro committed:
Add CODE_OF_CONDUCT.md

panosfol committed:
log: append `/` to log filepath

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
log: refactor the creation and usage of log file

This commit contains the necessary changes to have each database
implementation create and handle its own log file instead of a global log file.
This also makes it easier for threads to execute the db operations since we
don't have a single log file that needs syncronization

Changelog:

  - Removed the global variable `FILE *` for the log file. Each database
    implementation is now responsible for creating and handling each
    own log file, with its own name.

  - All the `pr_*_fd` macros now take the specific log file as argument
    since each database implementation will have a seperate log file.
    Some functions also needed changes to pass the log file as argument
    to ensure proper logging for each different log file.

  - Changed `struct db_t` to hold information about the log file name
    the log file, since each database handles its own file.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol committed:
db: add thread functionality

This commit utilizes the pthread.h library to change the flow of
execution of db operations with each database implementation
being executed in a different thread.

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

charmitro committed:
rust: email: cargo update

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>

panosfol committed:
cnc: fix help message alignment

Signed-off-by: Panagiotis Foliadis <pfoliadis@hotmail.com>

panosfol commented:
Convert to draft since there are a few errors indicated by valgind/helgrind.
Waiting for input from @charmitro .

charmitro reviewed (approved):
I generally approve this but we need to find a way to make the rust email library thread-safe also.

charmitro commented on configs/test.ini:
We still don't have a seperated config for this. I would also like to see a directory called `configs/samples/` where multiple configs for testing are to be stored there.

panosfol replied:
What would be the difference of the configs for testing from the configs that are already at `configs/`?

charmitro replied:
Forget the *samples* keyword, use `tests` instead.

* `configs/example.ini` - an example config with or without predefined values
* `configs/tests/<filename>` - the rest we use for testing.

panosfol replied:
Also, is there reason to believe our rust module is not safe in its current iteration?

charmitro commented on src/db.c:
```diff
@@ -37,26 +39,24 @@ void *db_operation_thread(void *arg)
 	init_db_func init_function;
 
+	// This is needed to access the localtime in a thread-safe manner
+	tzset();
```
Ping.

charmitro commented:
I'm getting weird helgrind output when running with email enabled, can 
you enable it properly (actually send an email) and see hellgrind's output?

panosfol commented:
> I'm getting weird helgrind output when running with email enabled, can you enable it properly (actually send an email) and see hellgrind's output?

Yes there are actually a lot of errors, all of them from the rust module. Also the actually sending of the email does not always happen for all the databases that were replicated, indicating that there is a conflict on the threads, making the outcome unpredictable.

charmitro commented:
> > I'm getting weird helgrind output when running with email enabled, can you enable it properly (actually send an email) and see hellgrind's output?
> 
> Yes there are actually a lot of errors, all of them from the rust module. Also the actually sending of the email does not always happen for all the databases that were replicated, indicating that there is a conflict on the threads, making the outcome unpredictable.

How should we handle this?
1. Should we postpone the merging of this PR until fixed? 
2. Disable email for more than one databases defined in the config?

panosfol commented:
> > > I'm getting weird helgrind output when running with email enabled, can you enable it properly (actually send an email) and see hellgrind's output?
> > 
> > 
> > Yes there are actually a lot of errors, all of them from the rust module. Also the actually sending of the email does not always happen for all the databases that were replicated, indicating that there is a conflict on the threads, making the outcome unpredictable.
> 
> How should we handle this?
> 
> 1. Should we postpone the merging of this PR until fixed?
> 2. Disable email for more than one databases defined in the config?

Lets postpone merging this, and look into multithreading in the rust library.
Also we probably should create a seperate issue for this, since its about the rust library.
All branch off of this branch and start looking into it.

charmitro commented on src/db.c:
```diff
@@ -2,6 +2,7 @@
 #include "db/postgres.h"
 #include "log.h"
 
+#include <bits/pthreadtypes.h>
```
What's this for?

charmitro committed:
Platform-specific includes for macOS and non-macOS systems

Removed the unnecessary `PATH_MAX` definition in util.h.

Added conditional inclusion of <bits/pthreadtypes.h> in db.c for
non-macOS systems. Included necessary headers for PATH_MAX and
NAME_MAX in src/log.c, src/postgres.c, and src/util.c when compiling
on macOS to ensure compatibility across different platforms.

Signed-off-by: Charalampos Mitrodimas <charmitro@posteo.net>
